Added optional choice for hdfs clients

pietermarsman opened the pull request:
## Description
Added optional choice for hdfs clients. Previously only InsecureClient was supported. This PR adds KerberosClient and TokenClient. 

## Motivation and Context
Some hdfs work with kerberos authentication. Mine does... This adds support for those clusters. 

This fixes #2481.

## Have you tested this? If so, how?

I have tested the KerberosClient in practice and it works! (Thousand times faster than hadoopcli :) ) 

Not tested the TokenClient though, because I do not have access to a Hadoop cluster with token authentication.

Tarrasch commented on luigi/contrib/hdfs/webhdfs_client.py:
```diff
@@ -43,6 +43,9 @@ class webhdfs(luigi.Config):
                               description='Port for webhdfs')
     user = luigi.Parameter(default='', description='Defaults to $USER envvar',
                            config_path=dict(section='hdfs', name='user'))
+    client_type = luigi.Parameter(default='insecure',
```
Maybe use ChoiceParameter?

Tarrasch commented on luigi/contrib/hdfs/webhdfs_client.py:
```diff
@@ -73,8 +78,18 @@ def client(self):
+            import hdfs
+            return hdfs.TokenClient(url=self.url, token=self.token)
+        else:
+            raise ValueError("Error: Unknown client type specified in webhdfs client_type"
```
You could skip this if you use the ChoiceParameter.

dlstadther commented on luigi/contrib/hdfs/webhdfs_client.py:
```diff
@@ -43,6 +43,9 @@ class webhdfs(luigi.Config):
                            config_path=dict(section='hdfs', name='user'))
+    client_type = luigi.Parameter(default='insecure',
+                                  description='Type of client to use. One of insecure, kerberos or token')
+    token = luigi.Parameter(default='', description='Hadoop delegation token, only used when client_type="token"')
```
@daveFNbuck Added an `OptionalParameter`. Would that be applicable here?

pietermarsman commented:
Only adding KerberosClient now.

And I've added some really basic test. Don't know what your standards are, so if you have any specific test in mind let me know. 

Also changed Parameter to ChoiceParameter. This removed the need for the ValueError is the else clause, so now I made InsecureClient the default in the else clause.

In the next couple of days I will test this code in a couple of projects.

pietermarsman commented:
(Y)(Y)

pietermarsman commented:
👍 👍

dlstadther commented:
Thanks for reporting back @pietermarsman ! Glad to hear things functioning as expected
